xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +155,797 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
+	xdsServer, nodeID, _, xdsResolver, stop := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{
+		AllowResourceSubset: true,
+	})
+	defer stop()
+
+	nonExistantServers, stopNonExistant := startTestServiceBackends(t, 2)
+	servers, stop := startTestServiceBackends(t, 2)
+	defer stop()
+
+	// Stop the servers that we want to be unreachable.
+	stopNonExistant()
+
+	newCluster1Name := "new_cluster_1"
+	newEdsService1Name := "new_eds_service_1"
+	newCluster2Name := "new_cluster_2"
+	newEdsService2Name := "new_eds_service_2"
+	clusterName := "aggregate_cluster"
+
+	ep1 := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: newEdsService1Name,
+		Localities: []e2e.LocalityOptions{{
+			Name:     "locality0",
+			Weight:   1,
+			Backends: backendOptions(t, nonExistantServers),
+		}},
+	})
+	ep2 := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: newEdsService2Name,
+		Localities: []e2e.LocalityOptions{{
+			Name:     "locality0",
+			Weight:   1,
+			Backends: backendOptions(t, servers),
+		}},
+	})
+	cluster1 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster1Name,
+		ServiceName: newEdsService1Name,
+	})
+	cluster2 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster2Name,
+		ServiceName: newEdsService2Name,
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		Type:        e2e.ClusterTypeAggregate,
+		// TODO: when "A75: xDS Aggregate Cluster Behavior Fixes" is merged, the
+		// policy will have to be set on the child clusters.
+		Policy:     e2e.LoadBalancingPolicyRingHash,
+		ChildNames: []string{newCluster1Name, newCluster2Name},
+	})
+	route := channelIDHashRoute(clusterName)
+	listener := e2e.DefaultClientListener("test.server", route.Name)
+
+	err := xdsServer.Update(context.Background(), e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{ep1, ep2},
+		Clusters:  []*v3clusterpb.Cluster{cluster, cluster1, cluster2},
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	gotPerBackend := checkRPCSendOK(t, ctx, client, 100)
+	if len(gotPerBackend) != 1 {
+		t.Errorf("Got RPCs routed to %v backends, want %v", len(gotPerBackend), 1)
+	}
+	for _, count := range gotPerBackend {
+		if count != 100 {
+			t.Errorf("Got %v RPCs routed to a backend, want %v", count, 100)
+		}
+	}
```

Comment:
Can we tighten this check?

At this point, it just verifies that all RPCs are sent to one backend. It does not verify that it is actually sent to *one* of the secondary backends. Although, I do agree that adding aggregate clusters into the mix here for ring_hash testing seems excessive. But given that we are following what C-core has in terms of test cases, we should actually verify that RPCs in fact are routed to the backends in the cluster that we expect RPCs to go to.

Reply:
I checked that we address the other backend, I had taken that from the c-core tests. I'm keeping a list of mistakes in the c-core tests, I'll open a ticket in grpc/grpc when I'm done.